NOAA-EMC/EVS
# HREF_restart

The PR for HREF_restart  has following updates:

(1) The MET/METplus  versions v12/v6 are used
(2) The restart capabilities are added to both stats and plots jobs
(3) The split_by_subset is applied in processing GDAS prepbufr data files
(4) The unit error in the title of the BSS plot for  T > 0C at 850mb is fixed
(5) Thoes levels with only single number of samples in the profile plots are removed
(6) The gather process is run in mpi parallel 
(7) The cat METplus log files to output o files are  removed from all 3 stat generation jobs
(8) Fixed the CPU over-subscription WARNING in the spcoutlook stat generation job 

Note: Using split_by_subset and running gather in mpi parallel together reduce the
      walltime from 4:30 to 3:30 hours for grid2obs, from 1:40 to 1:20 for precip and
      from 40min to 35 min for spcoutlook, respectively

Note:  If testing on personal account,  the following line should be added to the spcoutlook stat driver script: 
    export EVSINspcotlk=/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v2.0/prep/cam
or 
    export EVSINspcotlk=/lfs/h1/ops/prod/com/evs/v1.0/prep/cam 


Test procedures:

**Part 1.** For the stats generation jobs

There are 3 stats jobs:
 jevs_cam_href_grid2obs_stats.sh
 jevs_cam_href_precip_stats.sh
 jevs_cam_href_spcoutlook_stats.sh

For each job, following scenarios should be tested:

For jevs_cam_href_grid2obs_stats.sh:
**_Scenario 1:_** first run, no any interruption
In this case, all 3 output stat files (final stat files) should be generated and copied to the final stat directory
$COMOUT/cam/href.$VDATE
In the mean while the restart directory "restart" should be created in the small stat directory
$COMOUT/cam/atmos.$VDATE/href/grid2obs, in which all small stat files are saved (same as old version)
for gather processing or for restart run
In the restart directory, there are 4 sub-directories:
prepare, product,  profile, and  system

In the prepare sub-directory, there 2 *completed files: gfs_prepbufr.completed and rap_prepbufr.completed, and
sub-direcotory prepbufr.20240714 in which prepared prebbufr netCDF files are stored and for restart

In the product, or profile or system  sub-directory, there are several *.completed files to indicate which sub-tasks are completed
If the grid2obs job is fully completed without interruption, the completed files for all of the sub-tasks should be presented in
these 3 sub-directories, respectively.

 **_Scenario 2:_** the prepare process and a only part of the stats generation are completed but other part of stats generation is not.
Suppose  all processes in profile and system are completed but the processes in product are not.
To simulate this scenario:
Step 1: delete all of the output from the scenario 1, including the final stat files in $COMOUT/cam/href.$VDATE,
the small stat files and restart sub-directory in the $COMOUT/cam/atmos.$VDATE/href/grid2obs

Step 2: submit the driver script jevs_cam_href_grid2obs_stats.sh

Step 3: waiting about 1 hour, and kill the grid2obs job

Step 4: re-submit the driver script jevs_cam_href_grid2obs_stats.sh

Step 5: After it is completed, check the output final stat files, they should be same as the same files from scenario 1:

Repeat the above procedures for precip and spcoutlook jobs

**Part 2** For the plot generation jobs

There are 15 jobs:
7 jobs for 31-day score plots, 7 jobs for 90-day score plots and 1 job for precip spatial map:

All of those 15 jobs are pretty faster, mot of them are less than 15i min, but still have restart capability except for spatial map job.

For each job, a restart sub-directory is created in the $COMOUT directory:
$COMOUT/atmos.$VDATE/restart, in which there are additional 2 sub-directories 31 and 90 (for 31-day and 90-day plots,respectively)
in either 31 or 90 sub-directory, there are 7 sub-directories for each of 31-day jobs:
        href_cape_plots
        href_ctc_plots
        href_ecnt_plots
        href_precip_plots
        href_profile_plots
        href_snowfall_plots
        href_spcoutlook_plots

In each of these directory, all of completed png files and their indicating files *.completed are stored.
In the restart run, each of the *completed will be checked, if it exists, its corresponding png file is copied to the
working directory, otherwise, it will be generated by the job in the working directory
After all of png, eitehr are generated from the job or copied from the restart directory, they are combined into a big
tar file and staved in the COMOUT/atmos.$DVATE directory

The testing procedures for those 15 jobs deparnd on their walltime

(1) Following jobs are less than 1 min:
      jevs_cam_href_grid2obs_cape_past31days_plots.sh
      jevs_cam_href_grid2obs_cape_past90days_plots.sh
      jevs_cam_href_grid2obs_ctc_past31days_plots.sh
      jevs_cam_href_grid2obs_ctc_past90days_plots.sh
      jevs_cam_href_precip_past31days_plots.sh
      jevs_cam_href_precip_past90days_plots.sh

The restart testing can be skipped, and just the normal runs are tested
Note, if testing the restart capability. The procedure is:
    Step 1: launch the driver script, ushc as jevs_cam_href_grid2obs_cape_past31days_plots.sh,
    Step 2: wait about 30 or 40 seconds. kill the job and re-run the job
    Step 3: Check the final tar file to see if it is ok or not

(2) Other jobs has walltime between 2 ~ 16 min:
      jevs_cam_href_profile_past31days_plots.sh
      jevs_cam_href_profile_past90days_plots.sh
      jevs_cam_href_grid2obs_ecnt_past31days_plots.sh
      jevs_cam_href_grid2obs_ecnt_past90days_plots.sh
      jevs_cam_href_snowfall_past31days_plots.sh
      jevs_cam_href_snowfall_past90days_plots.sh
      jevs_cam_href_spcoutlook_past31days_plots.sh
      jevs_cam_href_spcoutlook_past90days_plots.sh

The testing procedure is similar to the above but kill the job
after waiting 1 or 2 minutes

(3) jevs_cam_href_precip_spatial_plots.sh
    No restart capability in this job since it is less than 1 min
